### ISSUE TITLE
Unable to install any packages/plugins in current bundle

### ISSUE THREAD

**Issue by jni** [napari/napari#2965]
## 🐛 Bug

Unable to install any packages/plugins in current bundle.

## To Reproduce

1. Download recent bundle build: https://github.com/napari/napari/actions/runs/988280586
2. Try to install any plugins or other packages in plugins > Install/Remove plugins...
3. Get the error: `error: unrecognized arguments: --upgrade`

Latest version of pip has not actually removed `--upgrade` so it's unclear to me what is going on. For reference the code that is failing is here:

https://github.com/napari/napari/blob/3aef5a5e6ab6c4fb4a83202040f1234ee4aca90b/napari/_qt/dialogs/qt_plugin_dialog.py#L82-L93

**Comment by sofroniewn** [comment]
@goanpeca maybe you can take a look at this when you get a chance

**Comment by jni** [comment]
Ok, the bug goes all the way back to 0.4.7 (doesn't work). 0.4.6 works. That points to the ironically-named #2436 as the likely culprit. 😂

**Comment by jni** [comment]
I should say I'm running all this on macOS 10.15.7 Catalina.

**Comment by sofroniewn** [comment]
I also see same thing on macOS 10.15.7 Catalina with nightly build

**Comment by goanpeca** [comment]
What is the pip version that comes with the bundle 🤔  ?

**Comment by jni** [comment]
> What is the pip version that comes with the bundle 🤔 ?

Difficult to answer because of #2964 😂

**Comment by jni** [comment]
> That points to the ironically-named #2436 as the likely culprit. 😂

Having said that, I don't see anything there that would cause this.... Except it corresponds to a briefcase update... 😬 😬 😬

**Comment by jni** [comment]
@goanpeca Given the error message I don't think pip is even being called. Perhaps it is not even being packaged???

**Comment by goanpeca** [comment]
~How can I download the artifact?~

Nevermind, the browser was not scrolling down

**Comment by jni** [comment]
Anyway, for reference 😂 

(1) Each release has artifacts attached: https://github.com/napari/napari/releases . With releases you have to click on "Assets" to see them.

(2) We have a nightly/on-push-to-master action that builds them: https://github.com/napari/napari/actions/workflows/make_bundle.yml

**Comment by jni** [comment]
My hypothesis is that the briefcase update in #2436 stopped packaging pip. However I don't know how to test it right now...

**Comment by jni** [comment]
never mind: both 0.4.6 and 0.4.7 contain pip 21.0.1 in `$APP/Contents/Resources/app_packages`.

**Comment by goanpeca** [comment]
<img width="532" alt="Screen Shot 2021-06-30 at 10 46 17 PM" src="https://user-images.githubusercontent.com/3627835/124061725-04762a80-d9f5-11eb-9174-d2afcb5f2341.png">

**Comment by goanpeca** [comment]
The command that is running is this `/Applications/napari.app/Contents/MacOS/napari -m pip install --upgrade ...` so it is running napari not python.

so it seem `sys.executable` is pointing to `/Applications/napari.app/Contents/MacOS/napari` and not to the bundled python.

I do not see a `python` executable anywhere on the bundled app contents.

**Comment by jni** [comment]
Yeah we are not yet signing our bundles. If you have experience with that @goanpeca a fix would be much appreciated!

But as an aside: screw Apple and their Move to Trash button.

> The command that is running is this

how did you find that @goanpeca? Nice sleuthing! I wonder then if actually what changed is how napari executes Python scripts... @tlambert03 played with that a fair bit...

Can you check whether the same thing happens in 0.4.6?

**Comment by goanpeca** [comment]
I modified the bundled files adding a `QMessageBox` and ran that on the install method inside .

`/Applications/napari.app/Contents/Resources/app/napari/_qt/dialogs/qt_plugin_dialog.py`

```python
    def install(self, pkg_list: Sequence[str]):
        cmd = ['-m', 'pip', 'install', '--upgrade']
        if running_as_bundled_app() and sys.platform.startswith('linux'):
            cmd += [
                '--no-warn-script-location',
                '--prefix',
                user_plugin_dir(),
            ]

        from qtpy.QtWidgets import QMessageBox
        QMessageBox.information(None, "test", " ".join([self.process.program()] + cmd + list(pkg_list)))
        self.process.setArguments(cmd + list(pkg_list))
        if self._output_widget:
            self._output_widget.clear()
        self.process.start()
```

**Comment by goanpeca** [comment]
>But as an aside: screw Apple and their Move to Trash button.

😆 
>Can you check whether the same thing happens in 0.4.6?

Will do

**Comment by jni** [comment]
> I wonder then if actually what changed is how napari executes Python scripts... @tlambert03 played with that a fair bit...

looking at the [release notes](https://github.com/napari/napari/releases/tag/v0.4.7) doesn't turn up anything suspicious/related to this...

**Comment by goanpeca** [comment]
@jni  on 0.4.6 it looks like, so it may be a briefcase thing that changed.

<img width="532" alt="Screen Shot 2021-06-30 at 11 53 24 PM" src="https://user-images.githubusercontent.com/3627835/124066500-5d968c00-d9fe-11eb-8905-96e88a9be6e1.png">

There is no `support/bin/python` available on the 0.4.9 bundle .

**Comment by jni** [comment]
Ok, so it was the briefcase upgrade after all. Hmm. That'll be fun to fix. LOL

**Comment by goanpeca** [comment]
In the meantime we should bin to the latest working version of it ?

**Comment by jni** [comment]
good Q. I think the update was required to e.g. have the version not be 0.0.1 in the bundle (see what happens when you open 0.4.6). Maybe there is an alternate strategy to achieve what #2436 was doing. CC @ziyangczi

**Comment by goanpeca** [comment]
@jni I asked Russell over the discord beeware channel

He said the following:

>Yes - a month or two back we made a modification to the way we package Python apps to get around some limitations with sandboxing in Big Sur.

>Big Sur wasn't very happy with the general approach of using a shell script as the "app", launching python as a second executable

>So - we now package a bootstrap app that is very similar to what happens on iOS - a C program that doesn't do much more than "start the Python interpreter"

>The downside is that fork-based subprocesses don't work any more, because there isn't a python interpreter (by default), and sys.executable is now hard coded to start a specific app.
(threads still work - it's only forks that are a problem)

>Option 1: Modify the stub app so it behaves like a "normal" python executable when it's called with python-like arguments (like -m), but behaves like the existing stub when it is given no arguments.
That would be the most generic solution - and it would also mean that fork-based processing would work again, because sys.executable would be a viable Python interpreter.

>Option 2: A custom Python support package that includes a python executable. We're currently explicitly removing bin/python from the support package because it isn't needed when the stub exists; but if you include it, you should be able to call it.

>(This very use case was raised earlier today by SpencerNewton - so there might be an argument to include the python binary as part of the default support package, and strip it as part of briefcase, with that stripping being an option that is enabled by default, but could be disabled by a project like Napari)

>Option 3: would be to work out how to invoke pip as an API - I know that's something that PyPA strictly advises against, but... it might still work :slight_smile:

So we have something to work with :)

**Comment by tlambert03** [comment]
> Yeah we are not yet signing our bundles. If you have experience with that @goanpeca a fix would be much appreciated!

Quick reminder here, it's not that we're not signing bundles (we are signing bundles), it's that we aren't signing with a proper (paid) Apple developer id.  

Still waiting for czi to buy us one of those ;)

**Comment by tlambert03** [comment]
> @jni I asked Russell over the discord beeware channel

That explains it! (I was trying to find python last night and wondering where it went)

That's a big deal for us... and perhaps reason to reconsider conda constructor or a variant like what I played with early on https://github.com/tlambert03/conda-bundler

**Comment by jni** [comment]
@goanpeca thanks for the investigation and reaching out to Russell! It looks like Option 1 and 2 both require dev on the briefcase side and would take at least a week to get going. So for the immediate term perhaps we should either:

- revert #2436 and/or pin briefcase (ie I'm not sure how much of #2436 needs to be reverted), or
- Use pip as a library temporarily :grimacing:. For reference the links about this are [the official warning](https://pip.pypa.io/en/latest/user_guide/#using-pip-from-your-program), and [this SO answer](https://stackoverflow.com/a/15950647/224254), ie `from pip._internal import main as pip_main; pip_main(['install', package_name])` (`pip.main` has been removed.)

By the way, between Option 1 and 2 I prefer option 1 because I don't think we want the napari bundle to break if a plugin uses `multiprocessing`. Having said that, I wonder whether in Option 2 `sys.executable` can be patched to point to the `python` binary and that would fix multiprocessing.

**Comment by tlambert03** [comment]
> * I'm not sure how much of [Fix packaging #2436](https://github.com/napari/napari/pull/2436) needs to be reverted), or

from a quick look, the only consequential thing in #2436 is the `if: runner.os == 'macOS' run: python -m pip install briefcase>=0.3.5` line.  (everything else is mostly "while you're in there" kind of stuff.)  So I'd vote for reverting that, and pinning to an earlier version.  
I'm not opposed to option 2, but only as a strictly temporary workaround, and since it will likely take a little testing and doing to make it work, I'm not sure it's worth it for a temp wrokaround.

**Comment by jni** [comment]
@tlambert03 it looks like we would reopen #2389 if we revert #2436, but that is preferable to not being able to install plugins imho. Do we all agree?

**Comment by tlambert03** [comment]
yeah, that seems to be the tradeoff.  fine with me

**Comment by liu-ziyang** [comment]
> good Q. I think the update was required to e.g. have the version not be 0.0.1 in the bundle (see what happens when you open 0.4.6). Maybe there is an alternate strategy to achieve what #2436 was doing. CC @ziyangczi

at the time being, I didn't find viable alternatives after trying things I can think of. I can look into this again a bit later this week (currently busy with other things)

Also no objection pining down the version, I can try following up with briefcase again and see if they have been making any progress making the latest version working for us. Unless @goanpeca if you happen to know how can we address #2389 without changing the start script in the application?

**Comment by tlambert03** [comment]
as noted in #3097, pinning briefcase to 0.3.1 in https://github.com/napari/napari/pull/2980 did _not_ actually change anything about this problem, since the newer support package (with static libs) is still being used.

we'll need to pin the support package as well, and since [3.8-b5](https://github.com/beeware/Python-Apple-support/releases/tag/3.8-b5) states that it's the first version in static lib format, rather than binary format, we probably need to pin earlier (see [here](https://briefcase.readthedocs.io/en/latest/reference/configuration.html?highlight=support_package#support-revision)):

```toml
[tool.briefcase.app.napari.macOS]
# 3.8-b4 was the last macOS support package in binary format, rather than static lib
# for python 3.9 use 3.9-b1
support_revision = 'b4'
```

@goanpeca, do you have time to fix this?  see #3097 ... where this is still popping up

**Comment by goanpeca** [comment]
Pinging @jaimergp
